segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/rolling/RollingStorage.java

```diff
@@ -0,0 +1,747 @@
+
+import com.google.common.base.Preconditions;
+import io.pravega.common.Exceptions;
+import io.pravega.common.LoggerHelpers;
+import io.pravega.common.util.ByteArraySegment;
+import io.pravega.common.util.CollectionHelpers;
+import io.pravega.segmentstore.contracts.BadOffsetException;
+import io.pravega.segmentstore.contracts.SegmentProperties;
+import io.pravega.segmentstore.contracts.StreamSegmentException;
+import io.pravega.segmentstore.contracts.StreamSegmentExistsException;
+import io.pravega.segmentstore.contracts.StreamSegmentInformation;
+import io.pravega.segmentstore.contracts.StreamSegmentNotExistsException;
+import io.pravega.segmentstore.contracts.StreamSegmentSealedException;
+import io.pravega.segmentstore.contracts.StreamSegmentTruncatedException;
+import io.pravega.segmentstore.contracts.StreamingException;
+import io.pravega.segmentstore.storage.SegmentHandle;
+import io.pravega.segmentstore.storage.SegmentRollingPolicy;
+import io.pravega.segmentstore.storage.StorageNotPrimaryException;
+import io.pravega.segmentstore.storage.SyncStorage;
+import io.pravega.shared.segment.StreamSegmentNameUtils;
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+
+/**
+ * A layer on top of a general SyncStorage implementation that allows rolling Segments on a size-based policy and truncating
+ * them at various offsets.
+ *
+ * Every Segment that is created using this Storage is made up of a Header and zero or more SegmentChunks.
+ * * The Header contains the Segment's Rolling Policy, as well as an ordered list of Offset-to-SegmentChunk pointers for
+ * all the SegmentChunks in the Segment.
+ * * The SegmentChunks contain data that their Segment is made of. A SegmentChunk starting at offset N with length L contains
+ * data for offsets [N,N+L) of the Segment.
+ * * A Segment is considered to exist if it has a non-empty Header and if its last SegmentChunk exists. If it does not have
+ * any SegmentChunks (freshly created), it is considered to exist.
+ * * A Segment is considered to be Sealed if its Header is sealed.
+ *
+ * A note about compatibility:
+ * * The RollingStorage wrapper is fully compatible with data and Segments that were created before RollingStorage was
+ * applied. That means that it can access and modify existing Segments that were created without a Header, but all new
+ * Segments will have a Header. As such, there is no need to do any sort of migration when starting to use this class.
+ * * Should the RollingStorage need to be discontinued without having to do a migration:
+ * ** The create() method should be overridden (in a derived class) to create new Segments natively (without a Header).
```

Comment:
To clarify, `in a derived class` means derived from `RollingStorage`, yes?

Reply:
yes.